Add KEP 1923 - try nominated node first (#2026)
kubernetes/enhancements: keps/sig-scheduling/1923-try-nominated-node-first/README.md

### Goals

In the case where `pod.Status.NominatedNodeName` is set for an incoming pod, the scheduler will evaluate the
nominated node first; if the nominated node doesn't fit the pod, the scheduling cycle will continue to evaluate
the rest of the nodes in the cluster just like we do today.

## Proposal

### User Stories (Optional)

Users want faster scheduling. Since it is highly likely the pod will only fit on the nominated node, the improvement
in scheduling latency will come at negligible cost (the cost being placing the pod on a less optimal node).

### Notes/Constraints/Caveats (Optional)

When this feature is enabled the preemptor Pod might not be dispatched to the best candidated node in some corner case,
e.g. another node releases the resources and becomes the best candidate while the victim pods got removed from the
nominated node.

## Design Details

### Implementation Details

1. In filtering phase, which is currently implemented in the method of `findNodesThatFitPod`, check the nominated node
   first if the incoming pod has the `pod.Status.NominatedNodeName` defined and the feature gate is enabled.

> 2. In case the nominated node doesn't suit for the incoming pod anymore, return `err` got from `findNodesThatPassFilters`,

Reviewer: I'm not sure I follow: If an error is returned, are you still suggesting to fail this scheduling cycle?

Author: not fail the scheduling cycle, continue to try other nodes. 
the error here is the we got from `findNodesThatFitPod` which only try the nominated node firstly, and it will continue to try other nodes on failure.

- [x] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
- [ ] (R) KEP approvers have approved the KEP status as `implementable`
- [x] (R) Design details are appropriately documented
- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
- [ ] (R) Graduation criteria is in place
- [ ] (R) Production readiness review completed
- [ ] Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

> If the scheduler fails to fit an incoming pod on any node, the scheduler will try to preempt lower

Author: Done

priority pods running on a selected node to make room for the pod. The name of this node will be set

> in the pod's `pod.Status.NominatedNodeName`.

Author: Done

The Node is called *Nominated* to indicate the intent for the Pod to be scheduled on it once preemption

> of other Pods finish. However, the `Pod.status.nominatedNodeName` information is not directly used in

Author: Done

the Pod's following scheduling attempts.

Pod scheduling is split into two phases, the scheduling cycle and the binding cycle, the scheduling cycle
primarily includes filtering and scoring.

When preemption happens in a previous scheduling cycle, there is a high chance that the nominated node is
the *only* node that satisfies the filters for the unscheduled Pod that triggered preemption.

This KEP proposes to change the scheduling cycle such that nominated node of a pod is evaluated first
and schedule the pod on that node if it fits. If the nominated node doesn't fit the pod, only then the

> scheduling cycle continues with the standard logic of evaluating the rest of the nodes in the cluster.

Author: Done

## Motivation

In real production environment, pods can have different priorites due to business needs, the preemption
could happen to make sure higher priority pods could get scheduled.

In cluster with large number of computing nodes, evaluating all nodes when scheduling a pod is time consuming.

### Goals

> In the case where `pod.Status.NominatedNodeName` is set for an incoming pod, the scheduler will evaluate the

Author: combined this suggestion and previous suggestion together.

nominated node first; if the nominated node doesn't fit the pod, the scheduling cycle will continue to evaluate
the rest of the nodes in the cluster just like we do today.

## Proposal

### User Stories (Optional)

Users want faster scheduling. Since it is highly likely the pod will only fit on the nominated node, the improvement
in scheduling latency will come at negligible cost (the cost being placing the pod on a less optimal node).

### Notes/Constraints/Caveats (Optional)

When this feature is enabled the preemptor Pod might not be dispatched to the best candidated node in some corner case,
e.g. another node releases the resources and becomes the best candidate while the victim pods got removed from the
nominated node.

## Design Details

### Implementation Details

1. In filtering phase, which is currently implemented in the method of `findNodesThatFitPod`, check the nominated node

>    first if the incoming pod has the `pod.Status.NominatedNodeName` defined and the feature gate is enabled.

Author: Done

> 2. In case the nominated node doesn't suit for the incoming pod anymore, get `err` from `findNodesThatPassFilters` where

Reviewer: Looks reasonable now.

However, I have one suggestion. Rather than evaluating the nominated first and only the rest of the nodes if it doesn't fit, we could do those 2 steps concurrently. Otherwise we would be increasing latency by potentially 2x in the worst case scenario.

Reviewer 2: why 2x? it is a single node evaluated through filters only.

Reviewer: If you have the number of Nodes equal to # of threads, it is 2x. But it's lower in general.

Reviewer 2: not worth the complexity in my opinion. it is not expensive to evaluate a single node through filters.

Reviewer: We still have to run PreFilter for all nodes initially, right?
In that case, yes, it might not be worth the complexity.

Reviewer 2: Right, I assumed prefilter would be run once.

Author: yes, prefilter will run only once, and the times to call those filters are not changed at all.

   `NominatedNode` is firstly evaluated, the `err` will be padded with more information to tell that scheduler is evaluating
   the feasibility of `NominatedNode` and failed on that node.

   If no error is returned but `NominatedNode` cannot pass all the filtering, this is possibly caused by the resource that
   claims to be removed but has not been fully released yet.

   For both of above cases, scheduler will continue to evaluate the rest of nodes to check if there is any node already
   available for the coming pod.

   Scheduler will retry until matching either of the following cases,
   - `NominatedNode` eventually released all the resource and the preemptor pod can be scheduled on that node.

>    - Another node in the cluster released enough release and pod get scheduled on that node instead.

Author: Done

     

### Test Plan

Following tests will be covered or considered:

- **Unit Tests**: All core changes must be covered by unit tests.
- **Integration Tests**: Integration test will be provided if necessary, for example,
  - enable the feature
  - preempt the victim pods on the nominated node
  - check pod will be scheduled on the nominated node
- **Benchmark Tests**: A benchmark test which compares the performance before and after the change.
  The performance improvement is visible by benchmark of `scheduling_algorithm_predicate_evaluation_seconds`.
  Other benchmark will be created on-demand along with the code review process.

### Graduation Criteria

#### Alpha (v1.21):

- [ ] New feature gate proposed to enable the feature.
- [ ] Implementation of the new feature in scheduling framework.
- [ ] Test cases mentioned in the [Test Plan](#test-plan).

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

_This section must be completed when targeting alpha to a release._

* **How can this feature be enabled / disabled in a live cluster?**
  - [x] Feature gate (also fill in values in `kep.yaml`)

>     - Feature gate name: TryNominatedNodeFirst

Author: Done

> # KEP-NNNN: Try Nominated Node First

Reviewer: 1923 :)

Author: DONE

<!-- toc -->
- [Release Signoff Checklist](#release-signoff-checklist)
- [Summary](#summary)
- [Motivation](#motivation)
  - [Goals](#goals)
- [Proposal](#proposal)
  - [User Stories (Optional)](#user-stories-optional)
  - [Notes/Constraints/Caveats (Optional)](#notesconstraintscaveats-optional)
- [Design Details](#design-details)
  - [Implementation Details](#implementation-details)
  - [Test Plan](#test-plan)
  - [Graduation Criteria](#graduation-criteria)
    - [Alpha (v1.21):](#alpha-v121)
- [Production Readiness Review Questionnaire](#production-readiness-review-questionnaire)
  - [Feature Enablement and Rollback](#feature-enablement-and-rollback)
- [Implementation History](#implementation-history)
<!-- /toc -->

## Release Signoff Checklist

Items marked with (R) are required *prior to targeting to a milestone / release*.

- [x] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)

> - [ ] (R) KEP approvers have approved the KEP status as `implementable`

Reviewer: You are already marking the KEP as implementable, so this should be checked. Or you could do that in a follow up PR if you want to have a first part merged first.

- [x] (R) Design details are appropriately documented
- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
- [ ] (R) Graduation criteria is in place
- [ ] (R) Production readiness review completed
- [ ] Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

If the incoming pod that going to be scheduled has the priority defined,

> and there is no available nodes in the cluster match with the needs for

Reviewer: ```suggestion
and there are no nodes in the cluster that satisfy the scheduling filters for the incoming Pod
```

the pod to be scheduled, scheduler will try to preempt lower priority pods

> running on each of the nodes in the cluster to release enough room for the

Reviewer: ```suggestion
running on a selected node in the cluster to make room for the
```

pod to be scheduled, the selected node for the incoming pod after the
preemption will be set as the value of `pod.Status.NominatedNodeName`.

> The nominated node here implies the node that is selected for the incoming

Reviewer: This sounds like it is the case today, but rather this is what the KEP is going to implement.

Reviewer: You have to state the problem:

"The Node is called Nominated to indicate the intent for the Pod to be scheduled on it once preemption of other Pods finish. However, the `Pod.status.nominatedNodeName` information is not directly used in the following scheduling cycles."

pod to be scheduled.

Pod scheduling is split into two phases, the scheduling cycle and the binding

> cycle, the scheduling cycle primarily includes predicate and scroing.

Reviewer: ```suggestion
cycle, the scheduling cycle primarily includes filtering and scoring.
```

> When the preemption is happened, the phase of predicate for the nominated node

Reviewer: When preemption happened in a previous scheduling cycle, there is a high chance that the nominated node is the *only* node that satisfies the filters for the incoming Pod to be scheduled.

has been properly measured, there is a high chance the selected nominated node
is still the *only* one candidate for the incoming pod to be scheduled.

> This KEP proposes to validate the availability of the nominated node directly

Reviewer: ```suggestion
This KEP proposes to validate the availability of the nominated node
```

in the new scheduling cycle before looping all over the nodes in the cluster
for the phase of the predicate.

## Motivation

In real production environment, pods can have different priorites due to business
needs, the preemption could happen to make sure high priority pods could got scheduled
and spawned.

For the cluster with huge number of computing nodes, looping over all the nodes in
the cluster for the phase of the predicate is an unnegligible consuming for scheduler,

> besides, it's unpredicatable that how the out-of-tree plugins are defined, whether

Reviewer: we don't need to mention out-of-tree plugins, the previous phrase should be enough.

they are time consuming, this is also a factor that may increase the time for the
phase of predicate.

### Goals

- In the case of `pod.Status.NominatedNodeName` is set for the incoming pod, scheduler

>   will try the nominated node first for the phase of predicate.

Reviewer: filtering phase.

> - Failed to schedule on the nominated node will fall back to verify all the nodes in

Reviewer: this shouldn't be a separate goal. Phrase it as part of the previous point.

  the cluster instead of preemption again.

> - A new feature gate will be provided for the operator to opt in/out this feature.

Reviewer: While in beta, but that's normal for any feature. Not sure if we should mention it.

Author: removed

## Proposal

### User Stories (Optional)

As a premium user, I want my high priority application got spin up more quickly,
I don't care where the application is running, whether there is any other nodes
in the cluster have more disk space, memory size etc., as long as my application
run normally.

### Notes/Constraints/Caveats (Optional)

> Enables this feature also implies the preemptor pod might not be dispatched

Reviewer: "When this feature is enabled the preemptor Pod"

to the best candidated node in some corner case, e.g. another node releases
the resources and becomes the best candidate while the victim pods got removed
from the nominated node, but enable this feature will make scheduler to consider

> the nominated node firstly.

Reviewer: the last sentence seem unnecessary.

## Design Details

### Implementation Details

> 1. In the phase of predicate, which is currently implemented in the method of `findNodesThatFitPod`,

Reviewer: filtering phase :)

   check the nominated directly if the incoming pod has the `pod.Status.NominatedNodeName` defined
   and the feature gate is enabled.

> 2. In case the nominated node is not suit the incoming pod anymore, return `fmt.Errorf` instead of

Reviewer: This shouldn't be like that.

Let's say the nominated Node had 2 Pods preempted and one was removed. At this point, kube-scheduler might try to schedule the Pod and fail. But in this case we should keep trying.

Also, errors are supposed to signal internal errors, meaning that something went wrong internally. This doesn't qualify.

Reviewer: Additionally, if the Pod doesn't fit in the nominated Node, aren't we supposed to iterate over the rest of the Nodes? Please add those details.

Reviewer: Maybe we can go onto a preemption evaluation cycle again and try the nominated Node first. There is a chance that the same Node will still be the nominated one. But if not, it gets changed.

Author: Hi Aldo,  I seriously considered those approaches and add a section to explain why it is designed as this, pls take a look and like to see your feedback.

   `core.FitError`, because this will give scheduler a chance to clean up the nominated node from
   the pod and find a new node to schedule instead of preemption again (might already have during
   the period). i.e. a fresh new scheduling cycle will be started from here.
3. If the nominated node is not suit for the pod anymore, the scheduling failure will be recorded
   and the `updatePod` will be called, here we change the logic to update the pod as long as the
   parameter `nominatedNode` is different with what pod holds in `pod.Status.NominatedNodeName`.
   In this case, parameter `nominatedNode` is an empty string so that the nominated node will be
   cleaned from the pod and the pod will be moved to the active queue. It lets scheduler find another

>    place for the pod.

Reviewer: in the next scheduling cycle

### Test Plan

Following tests wll be covered or considered:

- **Unit Tests**: All core changes must be covered by unit tests.

> - **Integration Tests**: Integration test will be provided if necessary, not visible in current stage.

Reviewer: Do a verbal description of what the test cases are supposed to look like.

We need to test for one preempted pod and several preempted pods as well.

Reviewer: Yes, integration and benchmark.

Author: I went through the integration test, I don't think the integration test is needed, since I cannot see any difference of expected result when this feature is enabled. 
And `scheduler_perf` already has the performance testcase to cover this, which is `scheduling_algorithm_predicate_evaluation_seconds`. I am not sure more testcases are needed or not.
Anyway, I have some placeholder created in this section.

- [x] (R) Design details are appropriately documented
- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
- [ ] (R) Graduation criteria is in place
- [ ] (R) Production readiness review completed
- [ ] Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

If the scheduler fails to fit an incoming pod on any node, the scheduler will try to preempt lower
priority pods running on a selected node to make room for the pod. The name of this node will be set

> in the pods' `pod.Status.NominatedNodeName`.

Author: Done

candidated node in some corner case, e.g. another node releases the resources and
becomes the best candidate while the victim pods got removed from the nominated node.

## Design Details

This feature hopes to improve the performance with the minimal change on the existing scheduling
framework, scheduler will short-circuit to check the nominated node directly instead of looping
all the nodes in the cluster, and if the nominated node doesn't fit, a fresh new scheduling cycle
will be started to get an new node for scheduling.

In most situation, the new scheduling cycle is not needed, but we still got the case like the
topology changing, victim pods deletion is still on-the-fly in the following scheduling cycle.
Those cases will make the nominated node not fits for the coming pod anymore, and a new scheduling
cycle is needed in those case. This is no harm since it will eventually get convergent.

### Implementation Details

1. In filtering phase, which is currently implemented in the method of `findNodesThatFitPod`,
   check the nominated directly if the incoming pod has the `pod.Status.NominatedNodeName` defined
   and the feature gate is enabled.

> 2. In case the nominated node not suit the incoming pod anymore, return `ErrNominateNode` instead of

Reviewer: I'm still not convinced that a single scheduling retry will be enough to discard the current nominated node. Maybe it will work if we solve kubernetes/kubernetes#94009 first

But to be clear, you are saying that if there is a nominated Node, that is the only Node we try in the scheduling cycle. Correct?

Author: this is not a problem in the current proposal, so mark it as resolved.

   `core.FitError`, because this will give scheduler a chance to clean up the nominated node from
   the pod and find a new node to schedule instead of preemption again (might already have during
   the period). i.e. a fresh new scheduling cycle will be started from here.
   A new error `ErrNominateNode` should be defined to describe what's going wrong on the nominated node.

> 3. If the nominated node is not suit for the pod anymore, the scheduling failure will be recorded

Reviewer: The enumeration makes it feel like 3 happens after 2, but in reality 3 is giving more details on 2. I don't think we need the paragraph at all.

Author: Done

- [x] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
- [x] (R) KEP approvers have approved the KEP status as `implementable`
- [x] (R) Design details are appropriately documented
- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
- [ ] (R) Graduation criteria is in place
- [ ] (R) Production readiness review completed
- [ ] Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

> If the incoming pod that going to be scheduled has the priority defined,
> and there are no nodes in the cluster that satisfy the scheduling filters
> for the incoming Pod, scheduler will try to preempt lower priority pods
> running on a selected node in the cluster to make room for the
> pod to be scheduled, the selected node for the incoming pod after the
> preemption will be set as the value of `pod.Status.NominatedNodeName`.

Reviewer 2: ```suggestion
If the scheduler fails to fit an incoming pod on any node, the scheduler will try to preempt lower priority pods
running on a selected node to make room for the pod. The name of this node will be set in the pods' `pod.Status.NominatedNodeName`.
```

Author: Done

The Node is called *Nominated* to indicate the intent for the Pod to be
scheduled on it once preemption of other Pods finish. However, the
`Pod.status.nominatedNodeName` information is not directly used in the

> following scheduling cycles.

Reviewer 2: ```suggestion
Pod's following scheduling attempts.
```

Author: Done

Pod scheduling is split into two phases, the scheduling cycle and the binding
cycle, the scheduling cycle primarily includes filtering and scoring.

> When the preemption happened in a previous scheduling cycle, there is a high

Reviewer 2: ```suggestion
When preemption happens in a previous scheduling cycle, there is a high
```

Author: Done

chance that the nominated node is the *only* node that satisfies the filters

> for the incoming Pod to be scheduled.

Reviewer 2: ```suggestion
for the unscheduled Pod that triggered preemption.
```

Author: Done

> This KEP proposes to validate the availability of the nominated node in the
> new scheduling cycle before looping all over the nodes in the cluster for the
> phase of the filtering.

Reviewer 2: ```suggestion
This KEP proposes to change the scheduling cycle such that nominated node of a pod is evaluated first and schedule the pod on that node if it fits. If the nominated node doesn't fit the pod, only then the scheduling cycle continues with the standard logic of evaluating the rest of the nodes in the cluster.
```

Author: Done

## Motivation

In real production environment, pods can have different priorites due to business

> needs, the preemption could happen to make sure high priority pods could got scheduled
> and spawned.

Reviewer 2: ```suggestion
needs, preemption could happen to make sure higher priority pods get scheduled.
```

Author: Done

> For the cluster with huge number of computing nodes, looping over all the nodes in
> the cluster for the phase of the filtering is an unnegligible consuming for scheduler.

Reviewer 2: ```suggestion
In clusters with large number of nodes, evaluating all nodes when scheduling a pod is time consuming.
```

Author: Done

### Goals

> - In the case of `pod.Status.NominatedNodeName` is set for the incoming pod, scheduler
>   will try the nominated node first for the filtering phase, failed to schedule on the
>   nominated node will fall back to verify all the nodes in the cluster instead of preemption
>   again.

Reviewer 2: ```suggestion
- In the case where `pod.Status.NominatedNodeName` is set for an incoming pod, the scheduler
  will evaluate the nominated node first; if the nominated node doesn't fit the pod, the scheduling cycle will continue to evaluate the rest of the nodes in the cluster just like we do today.
```

Author: Done

## Proposal

### User Stories (Optional)

As a premium user, I want my high priority application got spin up more quickly,
I don't care where the application is running, whether there is any other nodes
in the cluster have more disk space, memory size etc., as long as my application

> run normally.

Reviewer 2: I would phrase this as follows: Users want faster scheduling. Since it is highly likely the pod will only fit on the nominated node, the improvement in scheduling latency will come at negligible cost (the cost being placing the pod on a less optimal node).

Author: Done

### Notes/Constraints/Caveats (Optional)

When this feature is enabled the preemptor Pod might not be dispatched to the best
candidated node in some corner case, e.g. another node releases the resources and
becomes the best candidate while the victim pods got removed from the nominated node.

## Design Details

This feature hopes to improve the performance with the minimal change on the existing scheduling
framework, scheduler will short-circuit to check the nominated node directly instead of looping
all the nodes in the cluster, and if the nominated node doesn't fit, a fresh new scheduling cycle
will be started to get an new node for scheduling.

In most situation, the new scheduling cycle is not needed, but we still got the case like the
topology changing, victim pods deletion is still on-the-fly in the following scheduling cycle.
Those cases will make the nominated node not fits for the coming pod anymore, and a new scheduling

> cycle is needed in those case. This is no harm since it will eventually get convergent.

Reviewer 2: I would simplify this as: 

"We will update the scheduling cycle such that nominated node of a pod is evaluated first and schedule the pod on that node if it fits. If the nominated node doesn't fit the pod, only then the scheduling cycle continues with the standard logic of evaluating the rest of the nodes in the cluster."

Reviewer 2: we already mentioned that two times in the summary and goals, I feel we repeated the same thing multiple times unnecessarily. So I think we can even remove this section completely and directly get into the implementation details.

Author: removed.

### Implementation Details

1. In filtering phase, which is currently implemented in the method of `findNodesThatFitPod`,

>    check the nominated directly if the incoming pod has the `pod.Status.NominatedNodeName` defined

Reviewer 2: ```suggestion
   check the nominated node first if the incoming pod has the `pod.Status.NominatedNodeName` defined
```

Author: Done

   and the feature gate is enabled.

2. In case the nominated node doesn't suit for the incoming pod anymore, return `ErrNominateNode`
   instead of `core.FitError`, because this will give scheduler a chance to clean up the nominated
   node from the pod and find a new node to schedule instead of preemption again (might already have
   another node available for scheduling during the period). A fresh new scheduling cycle will be

>    started later.

Reviewer 2: are you suggesting that the rest of the nodes will not be evaluated in the same cycle? if so, why?

Author: I changed my mind to evaluate the rest of the nodes, this paragraph will be updated shortly.

Reviewer 2: thanks, was this updated? I can't tell because the new patch was an amend.

Reviewer: Keep in mind that removing the nominated node on the first failure is a source for other race conditions. For that reason, I suggest that all nodes should be evaluated and "new" nominated node calculated in the same cycle (if pod still doesn't fit).

Reviewer 2: right, and I am not sure we need to clear the nominatednodename if we successfully bind the pod to a node. We need to clear it only if we fail to find a node that fits and preemption fails to suggest a new one.

Author: sorry for delay, this is not updated by far, I am still thinking about @alculquicondor 's suggestion here.  I am inclined to follow his suggestion now.

Author: I am going to keep the `nominatedNode` untouched, there is a potential issue around the  `nominatedNode` , please see: https://github.com/kubernetes/kubernetes/issues/95752, this is the issue that we have today, and should be a separated topic instead.  I tracked the issue there and assign the issue to myself, so that we won't mess those issues together.
